### PR TITLE
Observable State

### PR DESCRIPTION
**Pull request by zilm13**

### CODE REVIEW

File: chain/src/main/java/org/ethereum/beacon/chain/observer/ObservableStateProcessorImpl.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      SpecHelpers specHelpers) {
+    this.chainStorage = chainStorage;
+    this.tupleStorage = chainStorage.getBeaconTupleStorage();
+    this.slotStatesStream = slotStatesStream;
+    this.specHelpers = specHelpers;
+    Flux.from(slotStatesStream).doOnNext(this::onSlotStateUpdate).subscribe();
+    Flux.from(pendingOperationsPublisher).doOnNext(this::onPendingStateUpdate).subscribe();
+  }
+
+  private void onSlotStateUpdate(BeaconState slotState) {
```
[the remark this answers is not in the document]
Comment (zilm13) [reply chain/src/main/java/org/ethereum/beacon/chain/observer/ObservableStateProcessorImpl.java]: I need it just to purge old attestation. I could use some `SlotTick` instead, but we don't have any concept like this yet.

**Comment by zilm13** [comment]
@mkalinin please, review

File: chain/src/main/java/org/ethereum/beacon/chain/observer/PendingOperationsState.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        .sorted(Comparator.comparing(Map.Entry::getKey))
+        .limit(maxCount)
+        .map(entry -> aggregateAttestations(entry.getValue()))
+        .collect(Collectors.toList());
+  }
+
+  private Attestation aggregateAttestations(List<Attestation> attestations) {
+    assert !attestations.isEmpty();
+
+    Bitfield participants =
```
[the remark this answers is not in the document]
Comment (zilm13) [reply chain/src/main/java/org/ethereum/beacon/chain/observer/PendingOperationsState.java]: I think this should be done outside there, immediately after receiving from network, so it's not covered by this class

File: chain/src/main/java/org/ethereum/beacon/chain/observer/PendingOperationsState.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        attestations.stream()
+            .map(Attestation::getParticipationBitfield)
+            .reduce(Bitfield::and)
+            .get();
+    Bitfield custody =
```
[the remark this answers is not in the document]
Comment (zilm13) [reply chain/src/main/java/org/ethereum/beacon/chain/observer/PendingOperationsState.java]: yeah but this code works wine both with zeros and real data, so I'd keep it here

File: chain/src/main/java/org/ethereum/beacon/chain/observer/ObservableStateProcessorImpl.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    Flux.from(attestationPublisher).doOnNext(this::onNewAttestation).subscribe();
+    Flux.from(beaconPublisher).doOnNext(this::onNewBlockTuple).subscribe();
+    this.executor =
+        Executors.newSingleThreadScheduledExecutor(
+            runnable -> {
+              Thread t = new Thread(runnable, "observable-state-processor");
+              t.setDaemon(true);
+              return t;
+            });
+    workIntervals.subscribe(tick -> executor.execute(this::doHardWork));
```
[the remark this answers is not in the document]
Comment (zilm13) [reply chain/src/main/java/org/ethereum/beacon/chain/observer/ObservableStateProcessorImpl.java]: done

File: chain/src/main/java/org/ethereum/beacon/chain/observer/ObservableStateProcessorImpl.java [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+  }
+
+  private void onNewSlot(SlotNumber newSlot) {
+    // From spec: Verify that attestation.data.slot <= state.slot - MIN_ATTESTATION_INCLUSION_DELAY
+    // < attestation.data.slot + EPOCH_LENGTH
+    // state.slot - MIN_ATTESTATION_INCLUSION_DELAY < attestation.data.slot + EPOCH_LENGTH
+    // state.slot - MIN_ATTESTATION_INCLUSION_DELAY - EPOCH_LENGTH < attestation.data.slot
+    purgeAttestations(
+        newSlot
```
[the remark this answers is not in the document]
Comment (zilm13) [reply chain/src/main/java/org/ethereum/beacon/chain/observer/ObservableStateProcessorImpl.java]: this condition is not defined anywhere in spec under some `methodName`, are you sure we should put such stuff to `SpecHelpers` ?

**Comment by zilm13** [comment]
ready for review @mkalinin

File: chain/src/main/java/org/ethereum/beacon/chain/observer/ObservableStateProcessorImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (mkalinin) [reply chain/src/main/java/org/ethereum/beacon/chain/observer/ObservableStateProcessorImpl.java]: Am not sure actually. It was a thought about potential issues that could happen later on.

**Pull request merged**